lib/futurism/resolver/controller/renderer.rb

```diff
@@ -0,0 +1,55 @@
+module Futurism
+  module Resolver
+    class Controller
+      class Renderer
+        def self.for(controller:, connection:, url:, params:)
+          new(controller: controller, connection: connection, url: url, params: params).renderer
+        end
+
+        def initialize(controller:, connection:, url:, params:)
+          @controller = controller
+          @connection = connection
+          @url = url || ""
+          @params = params || {}
+
+          setup_env!
+        end
+
+        def renderer
+          @renderer ||= controller.renderer
+        end
+
+        private
+
+        attr_reader :controller, :connection, :url, :params
+        attr_writer :renderer
+
+        def setup_env!
+          if url.present?
+            uri = URI.parse(url)
+            path = ActionDispatch::Journey::Router::Utils.normalize_path(uri.path)
+            query_hash = Rack::Utils.parse_nested_query(uri.query)
+
+            self.renderer =
+              renderer.new(
+                "rack.request.query_hash" => query_hash,
+                "rack.request.query_string" => uri.query,
+                "ORIGINAL_SCRIPT_NAME" => "",
+                "ORIGINAL_FULLPATH" => path,
+                Rack::SCRIPT_NAME => "",
+                Rack::PATH_INFO => path,
+                Rack::REQUEST_PATH => path,
+                Rack::QUERY_STRING => uri.query,
+                ActionDispatch::Http::Parameters::PARAMETERS_KEY => params.merge(query_hash)
```

Comment:
we're doing this to accommodate both form and query params, correct? Is there no simpler abstraction/factory method we could pluck from Rails source? I'm a bit fearful that this may become brittle if Rails' internals change in the future... I've shied away from using too much of Rails' private API in the past...

Reply:
I'm generally taking much of the heaving lifting from stimulus reflex, see https://github.com/hopsoft/stimulus_reflex/blob/9d3149cbe37f579420285938db13880214392001/lib/stimulus_reflex/reflex.rb#L64-L88 It's slightly different in that the SR code is for the reflex, whereas we're running this inside the channel. I more or less copied everything from that and fit it to meet our particular structure.

If you can find a clearer way, great, but that's what I was using as a basis.

Reply:
I did find some convenience methods that were all removed from rails along the way... so I think we'll leave this as it is. The only possible improvement I could see would be to extract some kind of factory method to make the code more succinct here. I don't insist on it though.